### ISSUE TITLE
ResponseCaching returns no headers in 304 NotModified response

### ISSUE THREAD

**Issue by mrjpeterj** [aspnet/AspNetCore#10583]
### Describe the bug
ResponseCaching returns no headers in 304 NotModified response

### To Reproduce
1. Setup a server that uses ResponseCaching
2. Make a request that responds with a Last-Modified and Expires header and see that it gets put into the ResponseCache
3. The Response will be a 200 with headers including Last-Modified, Expires, Cache-Control
3. Make the same request again, but add the If-Modified-Since header with a value the same as the previously returned Last-Modified time
4. The Response will be a 304 but it won't have any headers.

### Expected behavior
The 304 response should contain a number of headers as specified in RFC 7232

https://tools.ietf.org/html/rfc7232#section-4.1

```
   The server generating a 304 response MUST generate any of the
   following header fields that would have been sent in a 200 (OK)
   response to the same request: Cache-Control, Content-Location, Date,
   ETag, Expires, and Vary.
```

### Additional context

You can see this quite clearly in https://github.com/aspnet/AspNetCore/blob/master/src/Middleware/ResponseCaching/src/ResponseCachingMiddleware.cs#L156

**Comment by mikaelm12** [comment]
cc @JunTaoLuo

**Comment by anurse** [comment]
I can repro this and it does seem like it's contrary to the spec. @mrjpeterj is there a specific scenario that this affects or is this just about complying with the spec? I agree it is important to do so, but just trying to get an idea of the priority of this, if it's blocking a user scenario then it would be a higher priority.

**Comment by mrjpeterj** [comment]
Well it did crop up in implementing something for my job, but it turned out that for that specific thing I could just turn off server side caching to work around the issue. (setting Cache-Control: private)

The endpoint is using a MemoryCache to return the result anyway so it still performs fine.

If I was desperate for it, then I could implement the fix myself and submit it.

**Comment by anurse** [comment]
Ok, thanks for the context there! Like I said, I don't disagree it's a thing we should look at, but we likely won't get to it in 3.0 unless you (or someone else) is interested in submitting a PR. We'll review it again after 3.0 is done. Thanks for the report!

**Comment by alefranz** [comment]
I will give it a go

**Comment by alefranz** [comment]
This should have been fixed in #11428

**Comment by anurse** [comment]
Indeed! Unfortunately "Addresses" doesn't close the issue out when the PR is closed :).
